File: leet/0987-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from operator import itemgetter

class Solution:
    def verticalTraversal(self, root: TreeNode) -> List[List[int]]:
        if root is None:
            return []
			
        mm = {}
        self.dst(root, 0, 0, mm)
        ll = list(map(lambda y: y[1], sorted(mm.items(), key=lambda x: x[0])))
        return list(map(lambda y: self.getList(y), ll))
    
    def getList(self, ll: List) -> List[str]:
        return list(map(lambda x: x[1], sorted(ll, key=itemgetter(0,1))))
        
    
    def dst(self, root: TreeNode, curX: int, curY: int, mm: dict):
        if root is None:
            return
        
        if curX in mm:
            mm[curX].append((curY, root.val))
        else:
            mm[curX]=[(curY, root.val)]
            
        self.dst(root.left, curX - 1, curY + 1, mm)
        self.dst(root.right, curX + 1, curY + 1,mm)
        

File: leet/0987-vertical-order-traversal-of-a-binary-tree/vertical-order-traversal-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var verticalTraversal = function(root) {
    if (!root) return []
    
    let map = new Map()
    let ret = []
    
    let dfs = function(node, x, y){
        if(!node) return
        
        let arr = []
        if (map.has(x)) {
            arr = map.get(x)
        }else{
            map.set(x, arr)
        }
        
        arr.push({"val":node.val, "y":y})
        
        dfs(node.left, x - 1, y + 1)
        dfs(node.right, x + 1, y + 1)
    }
    
    dfs(root, 0, 0)
    
    let keys = []
    for (let key of map.keys()) {
        keys.push(key)
    }
    keys.sort((a, b) => a - b)
    
    for (let key of keys){
        let value = map.get(key)
        ret.push(value.sort((a, b) => {
            if (a.y == b.y) return a.val - b.val
            return a.y - b.y
        }).map(v => v.val))
    }
    
    return ret
};
